presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsCalculator.java

```diff
@@ -80,24 +85,36 @@ public PlanNodeStatsEstimate calculateStats(PlanNode node, StatsProvider sourceS
     }
 
     @Override
-    public void registerPlan(PlanNode root, Session session)
+    public boolean registerPlan(PlanNode root, Session session, long startTimeInNano, long timeoutInMilliseconds)
     {
         // Only precompute history based stats when plan has a join/aggregation.
-        if (!PlanNodeSearcher.searchFrom(root).where(node -> PRECOMPUTE_PLAN_NODES.stream().anyMatch(clazz -> clazz.isInstance(node))).matches()) {
-            return;
+        if (!prefetchForAllPlanNodes && !PlanNodeSearcher.searchFrom(root).where(node -> PRECOMPUTE_PLAN_NODES.stream().anyMatch(clazz -> clazz.isInstance(node))).matches()) {
+            return false;
         }
         ImmutableList.Builder<PlanNodeWithHash> planNodesWithHash = ImmutableList.builder();
-        forTree(PlanNode::getSources).depthFirstPreOrder(root).forEach(plan -> {
+        Iterable<PlanNode> planNodeIterable = forTree(PlanNode::getSources).depthFirstPreOrder(root);
+        for (PlanNode plan : planNodeIterable) {
+            if (checkTimeOut(startTimeInNano, timeoutInMilliseconds)) {
+                return false;
+            }
             if (plan.getStatsEquivalentPlanNode().isPresent()) {
                 planNodesWithHash.addAll(getPlanNodeHashes(plan, session).values());
             }
-        });
+        }
         try {
-            historyBasedStatisticsCacheManager.getStatisticsCache(session.getQueryId(), historyBasedPlanStatisticsProvider).getAll(planNodesWithHash.build());
+            historyBasedStatisticsCacheManager.getStatisticsCache(session.getQueryId(), historyBasedPlanStatisticsProvider, getHistoryBasedOptimizerTimeoutLimit(session)).getAll(planNodesWithHash.build());
         }
         catch (ExecutionException e) {
             throw new RuntimeException("Unable to register plan: ", e.getCause());
         }
+        // Return true even if get empty history statistics, so that HistoricalStatisticsEquivalentPlanMarkingOptimizer still return the plan with StatsEquivalentPlanNode which
+        // will be used in populating history statistics
+        return true;
```

Comment:
Even if fetch history timeouts or return empty, we still need StatsEquivalentPlanNode as it will be used in populating HBO history.